server/src/main/java/org/opensearch/transport/InboundBytesHandler.java

```diff
@@ -0,0 +1,21 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.transport;
+
+import org.opensearch.common.bytes.ReleasableBytesReference;
+
+import java.io.IOException;
+
+/**
+ * Interface for handling inbound bytes. Can be implemented by different transport protocols.
+ */
+public interface InboundBytesHandler {
+
+    public void doHandleBytes(TcpChannel channel, ReleasableBytesReference reference) throws IOException;
```

Comment:
It would useful to have `InboundBytesHandler` decide what it could handle, that would remove the need to have `TcpTransport::determineTransportProtocol`  altogether:

```
// For protobuf 
boolean canHandleBytes(BytesReference headerBuffer) {
            return (headerBuffer.get(0) == 'O' && headerBuffer.get(1) == 'S' && headerBuffer.get(2) == 'P') {
}

```

Reply:
So do you mean `InboundBytesHandler` has a method `canHandleBytes` that needs to be implemented by its implementations? It would still have to be decided which implementation of `InboundBytesHandler` to use, no?

Reply:
Correct, so each handler would check own header and InboundHandler would pick the one which handles the message (example here https://github.com/opensearch-project/OpenSearch/pull/12967#discussion_r1547977790)

Reply:
Got it! Though the determination for which bytes can be handled is not very simple for native protocol like it is for protobuf. Its quite complicated: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/transport/InboundDecoder.java#L80 as opposed to checking protobuf header prefix bytes. Thats why the determineTransportProtocol method is `return protobuf if (headerBuffer.get(0) == 'O' && headerBuffer.get(1) == 'S' && headerBuffer.get(2) == 'P')`.

Reply:
That would make it difficult to implement `canHandleBytes` for native  bytes handler. Do you think we can keep `determineTransportProtocol` for now and improve it later?

Reply:
The alternative solution would be to return `boolean` from `doHandleBytes`, if this case the native protocol would always win:

```
for (InboundMessageProtocolHandler handler: protocolBytesHandlers) {
    if (handler.doHandleBytes(reference) == true) {
       return;
}

throw new IllegalStateException("No bytes handler found for protocol: " + incomingMessageProtocol));
```

Reply:
Interesting! How would protobuf or any other protocol fit into that though? I feel `doHandleBytes` would have go through the reference to return something.

Reply:
This is what `determineTransportProtocol` does, right?

Reply:
Oh I thought you meant this would replace `determineTransportProtocol`. Never mind 🤦🏼‍♀️ 

So if I am understanding correctly, we keep the `determineTransportProtocol` and make `doHandleBytes` return a boolean?

Reply:
My apologies, I should have better have an example, this is how `doHandleBytes` would be implemented for Protobuf:

```
boolean doHandleBytes(ByteReference headerBuffer)  {
    if (headerBuffer.get(0) == 'O' && headerBuffer.get(1) == 'S' && headerBuffer.get(2) == 'P') {
       return true;
    } else {
       return false;
    }
```

For native, no change since we want it to go last. In this case, we at the moment have only one limitation - `native` protocol should always go last (it is easy to do), this snippet would basically go over all protocols in order and, if none found, finish up with `native`:

```
for (InboundMessageProtocolHandler handler: protocolBytesHandlers) {
    if (handler.doHandleBytes(reference) == true) {
       return;
    }
}

throw new IllegalStateException("No bytes handler found for protocol: " + incomingMessageProtocol));
```

Reply:
Ah that makes sense! So for `native`, it will always return true! Actually we can also have `canHandleBytes` always return true for native, since it will always go last in the list, we can make it return true. Other protocols like protobuf can have their conditions.

Reply:
Done! Let me know what you think.